# 2. Libraries and Functions

### Libraries

Python <b>libraries</b> (sometimes referred to as <i>packages</i>) are truly what make Python so useful. Because Python is free and open-source, anyone can make a Python library. Libraries consist of predefined functions and other capabilities. There are a few main libraries that we will discuss briefly here:

`math`: This library is built into Python itself, so you do not need to install it. It has various helpful features, such as a defined value for pi, as well as predefined functions for sine, cosine, tangent, etc. (`sin()`, `cos()`, etc.)

`numpy`: This is a NUMerical PYthon library, which has similar features as the `math` library, and many more. Numpy is one of the most commonly used libraries due to its extensive functionalities. We will discuss this one more in detail later on.

`scipy`: This is a SCIentific PYthon library, and has similar functions like `math` and `numpy`. Similarly, we will discuss this package in detail later on.

`matplotlib`: This is a plotting library that allows us to make graphs.

<b>To install a Python package</b>, we can do this in the command line. (You can search "terminal" on either Windows or Mac and it should pop up. It's a black screen.)

To install any package, run `pip install [package]` (for example, `pip install numpy`) in the command line. If it asks you for a yes/no input (usually y/n), say yes

- <b>Windows</b>: For more information, click [here](https://www.geeksforgeeks.org/how-to-install-numpy-on-windows/) for understanding how to install `numpy` (other packages can be installed similarly)
- <b>Mac</b>: For more information, click [here](https://www.geeksforgeeks.org/install-numpy-macos/) for understanding how to install `numpy` (other packages can be installed similarly)

To `import` a package (load it into Jupyter Notebook), we can run:

`import math as math`

What this is saying is that we are importing the `math` library, and we are calling it `math`. So, whenever we use a predefined function or variable, it will look like `math.[name]`. Like we said previously, the `math` library has a predefined value for pi. We can call upon this value by doing `math.pi`.

Let's try this now in the cell below:

In [1]:
import math as math

math.pi

3.141592653589793

Now, let's use this in an expression like we did in our first module (1_Basic_Math). We can insert `math.pi` into an expression just like any variable. Let's try it out by finding the area of a circle with a radius of two (`r=2`).

In [2]:
r = 2 # This is the radius

Area = math.pi * r**2 # This is our expression for Area, remember "**" is the exponent operation

Area

12.566370614359172

Now that we have our libraries sorted out and we know how to use them in code, we can start working on making our first function.

### Functions

<b>Functions</b> are an extremely important aspect of Python. They allow us to remember code snippets or formulas without defining them to a variable. Variables can be overwritten. Say you have `x` equal to one number, then later down in your code you redefine `x` as a different number. This will change the value of `x`. Let's show how this works: 

In [3]:
x = 2

print(x)

x = 3

2


Now, let's print `x` again and see what the answer is:

In [4]:
print(x)

3


We have changed the value! This was a simple example, but say we changed the value of `Area` as we had above, so that it was the area of a rectangle. This would get quite confusing. So, by using functions, we can ensure that we will always apply the proper formulas without them being overwritten. Let's write two functions: one for the area of a circle, and the other for the area of a rectangle.

In [5]:
def area_circle(r):
    return math.pi*r**2

Let's talk about what the above code means. We have a few different aspects:

`def` indicates that we are <i>defining</i> a function. This is a preset variable in Python (meaning you can't have something like `def=` in your code, it won't run!), and so it is green. 

`area_circle(r)` is the function with its parameter. If you remember from the last module, functions have the format `[name](parameters)`. This function only has one parameter, `r`. This is the radius of our circle.

`return` indicates what we will get out from the function. In this case, we want to get out the area of a circle, so we included the formula after the `return` (another built-in variable in Python, it's green!)

Now, let's see it in action with a circle of radius `3`:

In [6]:
area_circle(3)

28.274333882308138

Similar to our first area formula that we did where we had `r` defined before its use in the formula, we can do the same thing with our area function:

In [7]:
r = 3

area_circle(r)

28.274333882308138

And we get the same answer! Now let's look at the area of a rectangle, which we know is:

$$ A_{rectangle}=length*width $$

Let's call our length `x` and our width `y`:

In [8]:
def area_rectangle(x, y):
    area = x*y
    return area

We did something a bit different with this one: we first defined the area with the variable `area` inside of the function. Then, we included this newly defined variable in our `return`. This variable is defined <i>within</i> the function, so if we define `area` outside of the function, it doesn't affect the definition of our function. Let's try it out to see:

In [9]:
h = 2
b = 4

area = (1/2) * h * b # This is the area of a triangle

print(area)

# Let's use the same variables for the area of a rectangle:

area_rectangle(2, 4)

4.0


8

We can also define a function to a variable. This usually makes printing answers a bit simpler (`print(a)` versus `print(area_rectangle(2, 4))`, that second one is a mouthful and you might end up forgetting parentheses!)

In [10]:
a = area_rectangle(2, 4)

print(a)

8


So far, we have dealt with fairly simple equations that we turned into formulas. The second benefit of using functions is when formulas get more complex, like with our quadratic equation in the last module (1_Basic_Math). We don't want to risk that being overwritten, and I wouldn't want to have to write that again if it was. Let's make a function that describes the quadratic formula:

In [11]:
def quadform(a, b, c):
    ans1 = (-b + (b**2 - 4 * a * c)**(1/2)) / (2 * a)
    ans2 = (-b - (b**2 - 4 * a * c)**(1/2)) / (2 * a)
    return ans1, ans2

We can name our functions anything, but they follow the same rules as variables do. So, `quadform` and `QUADform` are two different functions. Usually, it is good to have your functions (and variables) in all lowercase, and I like to include `_`to represent spaces, like we did with the `area_rectangle` function. You also want your functions (and variables!) to be descriptive enough that you understand what they're doing (ex: `ar()` is much less descriptive than `area_rectangle()`).

Let's try out our new quadratic formula function:

In [12]:
quadform(1, 9, 2)

(-0.2279981273412348, -8.772001872658766)

Here, we get two outputs: addition in the numerator and subtraction, enclosed by parentheses.

### Practice

1. Write a function that describes the area of a triangle

2. Write a function for the volume of a cube

<b>Remember:</b> Make more cells with the "+" button (next to save icon). Move a cell up or down with the up and down arrows (next to "Run" button)

### Project 1: Split the Bill and Calculate Tip

Python is great for automating things in our everyday life. Have you ever gone to a restaurant and gotten one bill that has to be split between ten people and everyone pulls out their phones to start calculating how much they owe? Let's look at how we can split a bill evenly amongst friends, and calculate the tip as well.

In [15]:
bill = 243.89
amount_people = 12

Let's first define how we are going to solve for the tip. What do you need to figure out tip? The bill total, and the percent that you want to tip. This means our function will have two variables:

In [16]:
def tip(bill, percent):
    perc = percent / 100
    return bill * perc

In the above, we are having the `percent` parameter equal to a whole number between 0 and 100. Let's figure out how much our tip will be if we want to tip `20%`:

In [17]:
tip(bill, 20)

48.778

Now, let's do some arithmetic to add this to our bill to get our total bill:

In [18]:
bill_total = bill + tip(bill, 20)

print(bill_total)

292.668


Now that we have the total bill, let's split it between our group of diners. Let's make another function:

In [19]:
def split_bill(bill_total, number):
    return bill_total / number

And now we can split it:

In [20]:
split_bill(bill_total, amount_people)

24.389

Everyone owes $24.39, easy peasy. Because we defined these as functions, we can change the parameters to be any amount we would like and still compute the answers.

# When you're done with this notebook, make sure to save. Then, click "File", "Close and Halt" to exit